### PR TITLE
[SPARK-38176][SQL] ANSI mode: allow implicitly casting String to other simple types

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Compared to the default behavior, the current ANSI type coercion rules don't allow the following cases:

- Comparing String with other simple types, e.g. `str_col > date'2022-01-01'`
- Arithmetic operation containing String and other simple types
- Union/Intersect/Except containing String and other simple types
- SQL function expects non-string types but got  string input
- other SQL operators..

This PR is to remove the limitation. After changes, the String type can be implicit cast as Long/Double/Date/Timestamp/Boolean/Binary.

Note that Byte/Short/Int is not on the precedent list of String: `str_col > 1` will become `cast(str_col as long) > 1L`. So that we can avoid string parsing error if the string is out of the range of Byte/Short/Int in comparison/arithmetic/union operations.
The design applies to Float/Decimal (especially Decimal), for SQL operators containing Float/Decimal and String, the type coercion system will convert both as Double.
image

### Why are the changes needed?

The purpose of the current limitation is to prevent potential String parsing errors under ANSI mode. However, after doing research among real-world Spark SQL queries, I find that **many users are actually using String as Date/Timestamp/Numeric/etc in their queries**.  For example, the purpose of query `where str_col > date'2022-01-01'` is quite obvious, but users have to rewrite it as `where cast(str_col as date) > date'2022-01-01'` under ANSI mode.
To make the migration to ANSI mode easier, I suggest removing this limitation. Let's treat it as an extension in our SQL dialect.

### Does this PR introduce _any_ user-facing change?

Yes, allow implicitly casting String to other simple types under ANSI mode

### How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @cloud-fan @entong @allisonwang-db

File: docs/sql-ref-ansi-compliance.md [hunk 1]
```diff
@@ -188,34 +188,33 @@ java.lang.ArithmeticException: Casting 2147483648 to int causes overflow
...
+| Byte      | Byte -> Short -> Int -> Long -> Decimal -> Float* -> Double          |
+| Short     | Short -> Int -> Long -> Decimal-> Float* -> Double                   |
+| Int       | Int -> Long -> Decimal -> Float* -> Double                           |
+| Long      | Long -> Decimal -> Float* -> Double                                  |
+| Decimal   | Decimal -> Float* -> Double                                          |
+| Float     | Float -> Double                                                      |
+| Double    | Double                                                               |
+| Date      | Date -> Timestamp                                                    |
+| Timestamp | Timestamp                                                            |
+| String    | String, Long -> Double, Date -> Timestamp, Boolean, Interval, Binary |
```
Comment (cloud-fan) [inline docs/sql-ref-ansi-compliance.md:202]: I find this line a bit hard to understand. Shall we add some explanation below?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala [hunk 2]
```diff
@@ -171,17 +183,23 @@ object AnsiTypeCoercion extends TypeCoercionBase {
...
         Some(DoubleType)
 
-      // If the target type is any Decimal type, convert the String type literal as Double type.
-      case (StringType, DecimalType) if isInputFoldable =>
+      // If the target type is any Decimal type, convert the String type literal as the default
+      // Decimal type.
+      case (StringType, DecimalType) =>
         Some(DecimalType.SYSTEM_DEFAULT)
 
+      // If the target type is any Decimal type, convert the String type literal as the default
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:198]:
```suggestion
      // If the target type is any timestamp type, convert the String type literal as the default
```

**Review by cloud-fan — APPROVED** [review]

**Comment by gengliangwang** [comment]
Merging to master for unblocking https://issues.apache.org/jira/browse/SPARK-38154

**Pull request closed without merging**